url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/util/MersenneTwisterFast.java

```diff
@@ -0,0 +1,1304 @@
+                y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + (M - N)] ^ (y >>> 1) ^ mag01[y & 0x1];
+            }
+            y = (mt[N - 1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
+            mt[N - 1] = mt[M - 1] ^ (y >>> 1) ^ mag01[y & 0x1];
+
+            mti = 0;
+        }
+
+        y = mt[mti++];
+        y ^= y >>> 11;                          // TEMPERING_SHIFT_U(y)
+        y ^= (y << 7) & TEMPERING_MASK_B;       // TEMPERING_SHIFT_S(y)
+        y ^= (y << 15) & TEMPERING_MASK_C;      // TEMPERING_SHIFT_T(y)
+        y ^= (y >>> 18);                        // TEMPERING_SHIFT_L(y)
+
+        return (boolean) ((y >>> 31) != 0);
+    }
+
+    /** This generates a coin flip with a probability <tt>probability</tt>
+    of returning true, else returning false.  <tt>probability</tt> must
+    be between 0.0 and 1.0, inclusive.   Not as precise a random real
+    event as nextBoolean(double), but twice as fast. To explicitly
+    use this, remember you may need to cast to float first. */
+    public final boolean nextBoolean(final float probability) {
+        int y;
+
+        if (probability < 0.0f || probability > 1.0f) {
+            throw new IllegalArgumentException("probability must be between 0.0 and 1.0 inclusive.");
+        }
+        if (probability == 0.0f) {
+            return false;            // fix half-open issues
+        } else if (probability == 1.0f) {
+            return true;        // fix half-open issues
+        }
+        if (mti >= N) // generate N words at one time
+        {
+            int kk;
+            final int[] mt = this.mt; // locals are slightly faster
+            final int[] mag01 = this.mag01; // locals are slightly faster
+
+            for (kk = 0; kk < N - M; kk++) {
+                y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + M] ^ (y >>> 1) ^ mag01[y & 0x1];
+            }
+            for (; kk < N - 1; kk++) {
+                y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + (M - N)] ^ (y >>> 1) ^ mag01[y & 0x1];
+            }
+            y = (mt[N - 1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
+            mt[N - 1] = mt[M - 1] ^ (y >>> 1) ^ mag01[y & 0x1];
+
+            mti = 0;
+        }
+
+        y = mt[mti++];
+        y ^= y >>> 11;                          // TEMPERING_SHIFT_U(y)
+        y ^= (y << 7) & TEMPERING_MASK_B;       // TEMPERING_SHIFT_S(y)
+        y ^= (y << 15) & TEMPERING_MASK_C;      // TEMPERING_SHIFT_T(y)
+        y ^= (y >>> 18);                        // TEMPERING_SHIFT_L(y)
+
+        return (y >>> 8) / ((float) (1 << 24)) < probability;
+    }
+
+    /** This generates a coin flip with a probability <tt>probability</tt>
+    of returning true, else returning false.  <tt>probability</tt> must
+    be between 0.0 and 1.0, inclusive. */
+    public final boolean nextBoolean(final double probability) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 523-576 into a separate method.